### PR TITLE
Unified function signatures

### PR DESCRIPTION
**Pull request by b1llow**
# DO NOT SQUASH

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

An extract from https://github.com/rizinorg/rizin/pull/3463 that doesn't require updating Capstone.

**Test plan**

CI is green

### CODE REVIEW

**Comment by XVilka** [comment]
Seems like it needs some adjustment first - see broken unit test

**Comment by XVilka** [comment]
You will need also to remove the new Tricore tests in this PR

**Comment by XVilka** [comment]
Also:
```
gcc-12 -Ilibrz/bin/librz_bin.so.0.6.0.p -I. -I.. -Ilibrz -I../librz -Ilibrz/include -I../librz/include -I../librz/bin/format -I../subprojects/lz4-1.9.4/lib -Isubprojects/yxml -I../subprojects/yxml -Ilibrz/util/sdb/src -I../librz/util/sdb/src -Ilibrz/util/sdb -I../librz/util/sdb -I../librz/type/parser -I../subprojects/libmspack/libmspack/mspack -I../subprojects/libmspack/libmspack -fdiagnostics-color=always -fsanitize=address,undefined -fno-omit-frame-pointer -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Werror -O2 -g -Wimplicit-fallthrough=3 -DRZ_PLUGIN_INCORE=1 -D_GNU_SOURCE --std=gnu99 -Werror=sizeof-pointer-memaccess -fvisibility=hidden -DASAN=1 -DRZ_ASSERT_STDOUT=1 -Wno-cpp -ftrivial-auto-var-init=pattern -funsigned-char -fPIC -DRZ_API_BIN_ONLY=1 -MD -MQ librz/bin/librz_bin.so.0.6.0.p/bin.c.o -MF librz/bin/librz_bin.so.0.6.0.p/bin.c.o.d -o librz/bin/librz_bin.so.0.6.0.p/bin.c.o -c ../librz/bin/bin.c
../librz/bin/bin.c: In function ‘rz_bin_new’:
../librz/bin/bin.c:739:13: error: unused variable ‘i’ [-Werror=unused-variable]
  739 |         int i;
      |             ^
cc1: all warnings being treated as errors
```

**Comment by b1llow** [comment]
This test should still be related to memory

```
116/117 rizin:integration / open_analyse_save_load_project FAIL            1.60s   killed by signal 11 SIGSEGV
>>> MALLOC_PERTURB_=179 /private/var/folders/bc/yvvw2g2s4jlg9r8x7tf61xv800009d/T/woodpecker-local-693079402/rizinorg/rizin/build/test/integration/test_open_analyse_save_load_project
```

**Comment by XVilka** [comment]
@imbillow yes. Could not be a leak though, something like BOF or UAF. Try to run it with ASAN but leak detection disabled

File: librz/analysis/dwarf_process.c [hunk 1]
```diff
@@ -48,10 +48,10 @@ typedef struct dwarf_var_location_t {
...
 typedef enum variable_kind_t {
-	INVALID = 0,
-	FORMAL_PARAMETER,
-	VARIABLE,
-	UNSPECIFIED_PARAMETERS,
+	INVALID = RZ_ANALYSIS_VAR_KIND_INVALID,
+	FORMAL_PARAMETER = RZ_ANALYSIS_VAR_KIND_FORMAL_PARAMETER,
+	VARIABLE = RZ_ANALYSIS_VAR_KIND_VARIABLE,
+	UNSPECIFIED_PARAMETERS = RZ_ANALYSIS_VAR_KIND_UNSPECIFIED_PARAMETERS,
 } VariableKind;
```
Comment (wargio) [inline librz/analysis/dwarf_process.c]: If these are the same, why not use directly the same enum?

Comment (b1llow) [reply librz/analysis/dwarf_process.c]: You are right.

**Comment by XVilka** [comment]
Merged https://github.com/rizinorg/rizin/commit/94f2fae87ecd56cfaab5e782631a4a3858521a00, please rebase

**Review by XVilka — CHANGES_REQUESTED** [review]
Tests are fixed now but the unit test failure still here on macOS ARM64

File: librz/analysis/var.c [hunk 2]
```diff
@@ -722,14 +722,42 @@ static bool stack_offset_is_arg(RzAnalysisFunction *fcn, st64 stack_off) {
...
+	int n = rz_analysis_cc_max_arg(fcn->analysis, fcn->cc);
+	for (int i = 0; i < n; ++i) {
+		const char *x = rz_analysis_cc_arg(fcn->analysis, fcn->cc, i);
+		if (strcmp(x, regname) == 0)
+			return true;
+	}
+	return false;
+}
+
 RZ_API bool rz_analysis_var_is_arg(RzAnalysisVar *var) {
```
Comment (XVilka) [inline librz/analysis/var.c]: Could you please add Doxygen here while you are on it?

File: librz/analysis/var.c [hunk 3]
```diff
@@ -1314,15 +1342,37 @@ RZ_API void rz_analysis_fcn_vars_cache_fini(RzAnalysisFcnVarsCache *cache) {
...
+	char *highlighted_fcn_name = rz_str_newf("%s%s%s",
+		fcn_name_pre ? fcn_name_pre : "",
+		fcn_name,
+		fcn_name_post ? fcn_name_post : "");
+	char *highlighted_sig = rz_str_replace(sig, real_fcn_name, highlighted_fcn_name, false);
+	free(highlighted_fcn_name);
+	return highlighted_sig;
+}
+
 RZ_API char *rz_analysis_fcn_format_sig(RZ_NONNULL RzAnalysis *analysis, RZ_NONNULL RzAnalysisFunction *fcn, RZ_NULLABLE char *fcn_name,
```
Comment (XVilka) [inline librz/analysis/var.c:1365]: Same

**Review by XVilka — CHANGES_REQUESTED** [review]
Two more minor nitpicks

File: librz/analysis/var.c (hunk 2 above)
Comment (wargio) [reply librz/analysis/var.c]: also RZ_NONNULL

**Comment by wargio** [comment]
Looks like the `open_analyse_save_load_project` test is failing

File: librz/analysis/dwarf_process.c [hunk 4]
```diff
@@ -931,6 +989,8 @@ static const char *get_dwarf_reg_name(RZ_NONNULL char *arch, int reg_num, Variab
 		if (bits == 64) {
 			return map_dwarf_reg_to_ppc64_reg(reg_num, kind);
 		}
+	} else if (!strcmp(arch, "tricore")) {
+		return map_dwarf_reg_to_tricore_reg(reg_num, kind);
```
Comment (ret2libc) [inline librz/analysis/dwarf_process.c:993]: I wonder whether we should add a new `dwarf_reg_to_reg` method to the `RzAnalysisPlugin` instead to make this extensible and generic.

Comment (XVilka) [reply librz/analysis/dwarf_process.c:993]: Not in this PR please. But we could think about this, yes.

**Comment by XVilka** [comment]
@imbillow have you tried building with ASAN/UBSAN on macOS? Now that you have the hardware, should be straightforward.

File: librz/analysis/dwarf_process.c (hunk 4 above)
Comment (ret2libc) [reply librz/analysis/dwarf_process.c:993]: I agree, not in this PR. But it should be done if/before we are going to add another `if` here.

**Comment by b1llow** [comment]
> @imbillow have you tried building with ASAN/UBSAN on macOS? Now that you have the hardware, should be straightforward.

Yes, I built on macos using asan and fixed a lot of bugs and did some refactoring. But on the fix-leaks branch.

But maybe it's because my reconstruction seems to bring more problems.

File: librz/include/rz_analysis.h [hunk 5]
```diff
@@ -172,6 +172,7 @@ typedef struct rz_analysis_function_t {
 	bool is_pure;
 	bool is_variadic;
 	bool has_changed; // true if function may have changed since last anaysis TODO: set this attribute where necessary
+	bool has_debuginfo; ///< true if function has debug informations
```
Comment (XVilka) [inline librz/include/rz_analysis.h:175]: Just "information"

File: librz/core/cdwarf.c [hunk 6]
```diff
@@ -4,26 +4,30 @@
 #include <rz_core.h>
 
 RZ_API void rz_core_bin_dwarf_print_abbrev_section(const RzBinDwarfDebugAbbrev *da) {
```
Comment (XVilka) [inline librz/core/cdwarf.c]: Please add also `RZ_NULLABLE` here while you are on it, and Doxygen brief

**Review by XVilka — CHANGES_REQUESTED** [review]
Looks like it caused even bigger breakage

**Review by XVilka — CHANGES_REQUESTED** [review]
Looks like it caused even bigger breakage

**Review by XVilka — CHANGES_REQUESTED** [review]
Looks like it caused even bigger breakage

**Comment by XVilka** [comment]
@imbillow yes, please do not try to do everything in one PR. Just addressing the particular crash is enough, so that we could merge this PR. Further improvements could be done in separate PRs.

File: librz/bin/dwarf.c [hunk 7]
```diff
@@ -1281,41 +1281,60 @@ static void abbrev_decl_free(RzBinDwarfAbbrevDecl *ad) {
...
 }
 
-RZ_API RzBinDwarfAbbrevDecl *rz_bin_dwarf_debug_abbrev_get(RZ_NONNULL const RzBinDwarfDebugAbbrev *da, size_t idx) {
+RZ_API RzBinDwarfAbbrevDecl *rz_bin_dwarf_abbrev_get(RZ_NONNULL const RzBinDwarfDebugAbbrev *da, size_t idx) {
 	rz_return_val_if_fail(da, NULL);
 	return rz_pvector_at(&da->decls, idx);
 }
 
-RZ_API RzBinDwarfAbbrevDecl *rz_bin_dwarf_debug_abbrev_by_offet(RZ_NONNULL const RzBinDwarfDebugAbbrev *da, size_t offset) {
+RZ_API RzBinDwarfAbbrevDecl *rz_bin_dwarf_abbrev_by_offet(RZ_NONNULL const RzBinDwarfDebugAbbrev *da, size_t offset) {
```
Comment (XVilka) [inline librz/bin/dwarf.c]: By the way, note that it should be `offset`, not `offet` which is obvious typo

**Comment by XVilka** [comment]
I recommend to extract these further commits into one more separate PR and focus on in this PR on fixing just one crash.

**Comment by b1llow** [comment]
It took me a lot of time to realize that not checking the return values of `ht_pp_insert` and `sdb_set_own` could lead to a memory leak.

**Comment by XVilka** [comment]
Now there is a problem in analysis_graph unit test

**Comment by XVilka** [comment]
@imbillow looks like it passed all tests on macOS. Please squash some commits then we can merge if green.

**Review by wargio — APPROVED** [review]

**Comment by b1llow** [comment]
I found that the build with ASAN does not detect the bug, but buildtype=release does.

**Comment by XVilka** [comment]
@imbillow yes, sometimes it happens. Also, as you can see, macOS has stricter runtime than Linux or Windows.

**Review by XVilka — APPROVED** [review]

**Pull request merged**